cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3431 @@
+      success &= GenerateTerminator(offset, rng, arr);
+      break;
+    case 3:
+      success &= GenerateNounPhrase(offset, rng, arr);
+      success &= GeneratePrepositionalPhrase(offset, rng, arr);
+      success &= GenerateVerbPhrase(offset, rng, arr);
+      success &= GenerateNounPhrase(offset, rng, arr);
+      success &= GenerateTerminator(offset, rng, arr);
+      break;
+    case 4:
+      success &= GenerateNounPhrase(offset, rng, arr);
+      success &= GeneratePrepositionalPhrase(offset, rng, arr);
+      success &= GenerateVerbPhrase(offset, rng, arr);
+      success &= GeneratePrepositionalPhrase(offset, rng, arr);
+      success &= GenerateTerminator(offset, rng, arr);
+      break;
+    default:
+      Unreachable("Random number should be between 0 and 5 inclusive");
+      break;
+  }
+  return success;
+}
+
+class TpchTableGenerator {
+ public:
+  using OutputBatchCallback = std::function<void(ExecBatch)>;
+  using FinishedCallback = std::function<void(int64_t)>;
+  using GenerateFn = std::function<Status(size_t)>;
+  using ScheduleCallback = std::function<Status(GenerateFn)>;
+  using AbortCallback = std::function<void()>;
+
+  virtual Status Init(std::vector<std::string> columns, float scale_factor,
+                      int64_t batch_size) = 0;
+
+  virtual Status StartProducing(size_t num_threads, OutputBatchCallback output_callback,
+                                FinishedCallback finished_callback,
+                                ScheduleCallback schedule_callback) = 0;
+
+  bool Abort() {
+    bool expected = false;
+    return done_.compare_exchange_strong(expected, true);
+  }
+
+  virtual std::shared_ptr<Schema> schema() const = 0;
+
+  virtual ~TpchTableGenerator() = default;
+
+ protected:
+  std::atomic<bool> done_ = {false};
+  std::atomic<int64_t> batches_outputted_ = {0};
+};
+
+int GetNumDigits(int64_t x) {
+  // This if statement chain is for MAXIMUM SPEED
+  // Source:
+  // https://stackoverflow.com/questions/1068849/how-do-i-determine-the-number-of-digits-of-an-integer-in-c
+  ARROW_DCHECK(x >= 0);
+  if (x < 10ll) return 1;
+  if (x < 100ll) return 2;
+  if (x < 1000ll) return 3;
+  if (x < 10000ll) return 4;
+  if (x < 100000ll) return 5;
+  if (x < 1000000ll) return 6;
+  if (x < 10000000ll) return 7;
+  if (x < 100000000ll) return 8;
+  if (x < 1000000000ll) return 9;
+  if (x < 10000000000ll) return 10;
+  if (x < 100000000000ll) return 11;
+  if (x < 1000000000000ll) return 12;
+  if (x < 10000000000000ll) return 13;
+  if (x < 100000000000000ll) return 14;
+  if (x < 1000000000000000ll) return 15;
+  if (x < 10000000000000000ll) return 16;
+  if (x < 100000000000000000ll) return 17;
+  if (x < 1000000000000000000ll) return 18;
+  return -1;
+}
+
+void AppendNumberPaddedToNineDigits(char* out, int64_t x) {
+  // We do all of this to avoid calling snprintf, which needs to handle locale,
+  // which can be slow, especially on Mac and Windows.
+  int num_digits = GetNumDigits(x);
```

Comment:
Why doesn't this use the existing formatting functionalities in `arrow/util/formatting.h`?